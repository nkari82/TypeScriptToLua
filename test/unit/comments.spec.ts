import * as util from "../util";

test("Single-line JSDoc is copied on a function", () => {
    const builder = util.testModule`
        /** This is a function comment. */
        function foo() {}
    `
        .expectToHaveNoDiagnostics()
        .expectDiagnosticsToMatchSnapshot();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("This is a function comment.");
});

test("Multi-line JSDoc with one block is copied on a function", () => {
    const builder = util.testModule`
        /**
         * This is a function comment.
         * It has more than one line.
         */
        function foo() {}
    `
        .expectToHaveNoDiagnostics()
        .expectDiagnosticsToMatchSnapshot();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("It has more than one line.");
});

test("Multi-line JSDoc with two blocks is copied on a function", () => {
    const builder = util.testModule`
        /**
         * This is a function comment.
         * It has more than one line.
         *
         * It also has more than one block.
         */
        function foo() {}
    `
        .expectToHaveNoDiagnostics()
        .expectDiagnosticsToMatchSnapshot();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("It also has more than one block.");
});

test("JSDoc is copied on a function with tags", () => {
    const builder = util.testModule`
        /**
         * This is a function comment.
         * It has multiple lines.
         *
         * @param arg1 This is the first argument.
         * @param arg2 This is the second argument.
         * @returns A very powerful string.
         */
        function foo(arg1: boolean, arg2: number): string {
            return "bar";
        }
    `
        .expectToHaveNoDiagnostics()
        .expectDiagnosticsToMatchSnapshot();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("This is the first argument.");
    expect(lua).toContain("This is the second argument.");
    expect(lua).toContain("A very powerful string.");
});

test("JSDoc is copied on a variable", () => {
    const builder = util.testModule`
        /** This is a variable comment. */
        const foo = 123;
    `
        .expectToHaveNoDiagnostics()
        .expectDiagnosticsToMatchSnapshot();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("This is a variable comment.");
});

// https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1299
test("tsdoc comment does not cause a diagnostic (#1299)", () => {
    util.testModule`
        interface LuaBootstrap {
            /**
             * @remarks
             *
             * Links can point to a URL: {@link https://github.com/microsoft/tsdoc}
             */
            on_init(f: (() => void) ): void
        }

        const script : LuaBootstrap = {
            on_init: (x) => x()
        }

        script.on_init(() => {})
    `.expectToHaveNoDiagnostics();
});
