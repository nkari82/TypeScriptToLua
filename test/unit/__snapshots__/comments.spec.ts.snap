// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSDoc is copied on a function with tags: code 1`] = `
"--- This is a function comment.
-- It has multiple lines.
-- 
-- @param arg1 This is the first argument.
-- @param arg2 This is the second argument.
-- @returns A very powerful string.
function foo(self, arg1, arg2)
    return "bar"
end"
`;

exports[`JSDoc is copied on a function with tags: diagnostics 1`] = `""`;

exports[`JSDoc is copied on a variable: code 1`] = `
"--- This is a variable comment.
foo = 123"
`;

exports[`JSDoc is copied on a variable: diagnostics 1`] = `""`;

exports[`Multi-line JSDoc with one block is copied on a function: code 1`] = `
"--- This is a function comment.
-- It has more than one line.
function foo(self)
end"
`;

exports[`Multi-line JSDoc with one block is copied on a function: diagnostics 1`] = `""`;

exports[`Multi-line JSDoc with two blocks is copied on a function: code 1`] = `
"--- This is a function comment.
-- It has more than one line.
-- 
-- It also has more than one block.
function foo(self)
end"
`;

exports[`Multi-line JSDoc with two blocks is copied on a function: diagnostics 1`] = `""`;

exports[`Single-line JSDoc is copied on a function: code 1`] = `
"--- This is a function comment.
function foo(self)
end"
`;

exports[`Single-line JSDoc is copied on a function: diagnostics 1`] = `""`;
